File: app/src/main/java/com/volkanhotur/basemvvm/android/data/domain/BaseObserver.java
package com.volkanhotur.basemvvm.android.data.domain;

import android.widget.Toast;

import com.core.abstractbase.AbstractView;
import com.domain.remote.exception.AbstractApiException;
import com.domain.remote.usecase.AbstractObserver;

/**
 * @author volkanhotur
 */

public abstract class BaseObserver<T> extends AbstractObserver<T> {

    private AbstractView view;

    protected BaseObserver(AbstractView view) {
        this.view = view;
    }

    protected BaseObserver() {
        this.view = null;
    }

    @Override
    protected void onStart() {
        super.onStart();
        if (view != null && view.isLoadingBarEnabled())
            view.showLoadingBar();
    }

    @Override
    public void onNext(T t) {
        super.onNext(t);
    }

    @Override
    public void onComplete() {
        super.onComplete();
        if (view != null && view.isLoadingBarEnabled())
            view.hideLoadingBar();
    }

    @Override
    public void onError(Throwable exception) {
        super.onError(exception);

        if (view != null && view.isLoadingBarEnabled())
            view.hideLoadingBar();

        if (view != null) {
            if(exception instanceof AbstractApiException){
                 if(((AbstractApiException) exception).getCode() == 401 || ((AbstractApiException) exception).getCode() == 403){
                    view.expireSession();
                }  else {
                     Toast.makeText(view.context(), "We are unable to process at this time. Please try again later.", Toast.LENGTH_LONG).show();
                 }
            }else {
                Toast.makeText(view.context(), "We are unable to process at this time. Please try again later.", Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/com/volkanhotur/basemvvm/android/data/domain/BaseObserver.kt
package com.volkanhotur.basemvvm.android.data.domain

import android.widget.Toast
import com.core.abstractbase.AbstractView
import com.domain.remote.exception.AbstractApiException
import com.domain.remote.usecase.AbstractObserver
import com.volkanhotur.basemvvm.R
import com.volkanhotur.basemvvm.android.extension.toastLong

/**
 * @author volkanhotur
 */
abstract class BaseObserver<T> : AbstractObserver<T> {

    private var view: AbstractView?

    protected constructor(view: AbstractView?) {
        this.view = view
    }

    protected constructor() {
        view = null
    }

    override fun onStart() {
        super.onStart()

        view?.let {
            if (it.isLoadingBarEnabled) {
                it.showLoadingBar()
            }
        }
    }

    override fun onNext(t: T) {
        super.onNext(t)
    }

    override fun onComplete() {
        super.onComplete()

        view?.let {
            if (it.isLoadingBarEnabled) {
                it.hideLoadingBar()
            }
        }
    }

    override fun onError(exception: Throwable) {
        super.onError(exception)

        view?.let {
            if (it.isLoadingBarEnabled) {
                it.hideLoadingBar()
            }

            if (exception is AbstractApiException) {
                if (exception.code == 401 || exception.code == 403) {
                    it.expireSession()
                } else {
                    it.context().toastLong(R.string.unknown_error)
                }
            } else {
                it.context().toastLong(R.string.unknown_error)
            }
        }
    }
}